Perform case insensitive lookups for validation rules, disposable domains

mtillberg opened the pull request:
Validation wasn't working on a few cases where email address domains had upper case letters, but lookups were exact matches against lower case strings.  This adds tests for and fixes requireOnlyTopLevelDomains, disallowReservedDomains, and FileSource domain lookups.

RohanNagar commented:
Thanks for your contribution @mtillberg! This does seem like a proper fix, though I wonder if there are any cases where case-sensitivity is desired? I'll have a little think about this.

RohanNagar commented on src/main/java/com/sanctionco/jmail/ValidationRules.java:
I agree that this is probably a better place for it if we want to apply case-insensitivity.

mtillberg commented:
TopLevelDomain now has case insensitive comparisons/hash codes.  Using compareToIgnoreCase is faster than equalsIgnoreCase or toLowerCase.  I also moved the toLowerCase into ValidationRules for disposable domains, and added a few comments about lowercase domains.  I updated the test case to show upper/mixed case domains don't match, more as a marker for behavior.

mtillberg commented on src/main/java/com/sanctionco/jmail/TopLevelDomain.java:
```diff
@@ -55,12 +55,14 @@ public boolean equals(Object o) {
-    return Objects.equals(tld, that.tld);
+    if (tld == null) return that.tld == null;
+    if (that.tld == null) return false;
+    return tld.compareToIgnoreCase(that.tld) == 0;
```
compareToIgnoreCase is faster than equalsIgnoreCase here (ObjectEquals is the old case sensitive way, EqualsLowerCase is .toLowerCase() on both sides):

```
StringComparisonBenchmark.testCompareIgnoreCase  thrpt    5  2012.449 ±  47.200  ops/ms
StringComparisonBenchmark.testEqualsIgnoreCase   thrpt    5  1451.934 ± 225.504  ops/ms
StringComparisonBenchmark.testEqualsLowerCase    thrpt    5   807.040 ±  27.133  ops/ms
StringComparisonBenchmark.testEqualsNew          thrpt    5  2002.633 ±  24.730  ops/ms
StringComparisonBenchmark.testObjectEquals       thrpt    5  4367.602 ±  58.549  ops/ms
```

The null checks are because you have EqualsVerifier in the TopLevelDomain tests.  That forces null handling, and has surprisingly little effect (EqualsNew is the new comparison implementation).

RohanNagar commented:
Makes sense, I agree with that. I'll likely do some refactoring after merging both of these changes.

RohanNagar commented:
Thank you for the great work @mtillberg!
